lib/internal/crypto/keys.js

```diff
@@ -0,0 +1,345 @@
+    return defaultFormat;
+  else if (formatStr === 'pem')
+    return kKeyFormatPEM;
+  else if (formatStr === 'der')
+    return kKeyFormatDER;
+  throw new ERR_INVALID_OPT_VALUE(optionName, formatStr);
+}
+
+function parseKeyType(typeStr, required, keyType, isPublic, optionName) {
+  if (typeStr === undefined && !required) {
+    return undefined;
+  } else if (typeStr === 'pkcs1') {
+    if (keyType !== undefined && keyType !== 'rsa') {
+      throw new ERR_CRYPTO_INCOMPATIBLE_KEY_OPTIONS(
+        typeStr, 'can only be used for RSA keys');
+    }
+    return kKeyEncodingPKCS1;
+  } else if (typeStr === 'spki' && isPublic !== false) {
+    return kKeyEncodingSPKI;
+  } else if (typeStr === 'pkcs8' && isPublic !== true) {
+    return kKeyEncodingPKCS8;
+  } else if (typeStr === 'sec1' && isPublic !== true) {
+    if (keyType !== undefined && keyType !== 'ec') {
+      throw new ERR_CRYPTO_INCOMPATIBLE_KEY_OPTIONS(
+        typeStr, 'can only be used for EC keys');
+    }
+    return kKeyEncodingSEC1;
+  }
+
+  throw new ERR_INVALID_OPT_VALUE(optionName, typeStr);
+}
+
+function option(name, objName) {
+  return objName === undefined ? name : `${objName}.${name}`;
+}
+
+function parseKeyFormatAndType(enc, keyType, isPublic, objName) {
+  const { format: formatStr, type: typeStr } = enc;
+
+  const isInput = keyType === undefined;
+  const format = parseKeyFormat(formatStr,
+                                isInput ? kKeyFormatPEM : undefined,
+                                option('format', objName));
+
+  const type = parseKeyType(typeStr,
+                            !isInput || format === kKeyFormatDER,
+                            keyType,
+                            isPublic,
+                            option('type', objName));
+
+  return { format, type };
+}
+
+function isStringOrBuffer(val) {
+  return typeof val === 'string' || isArrayBufferView(val);
+}
+
+function parseKeyEncoding(enc, keyType, isPublic, objName) {
+  const isInput = keyType === undefined;
+
+  const {
+    format,
+    type
+  } = parseKeyFormatAndType(enc, keyType, isPublic, objName);
+
+  let cipher, passphrase;
+  if (isPublic !== true) {
+    ({ cipher, passphrase } = enc);
+
+    if (!isInput && cipher != null) {
+      if (typeof cipher !== 'string')
+        throw new ERR_INVALID_OPT_VALUE(option('cipher', objName), cipher);
+      if (format === kKeyFormatDER &&
+          (type === kKeyEncodingPKCS1 ||
+           type === kKeyEncodingSEC1)) {
+        throw new ERR_CRYPTO_INCOMPATIBLE_KEY_OPTIONS(
+          encodingNames[type], 'does not support encryption');
+      }
+    }
+
+    if ((isInput && passphrase !== undefined &&
+         !isStringOrBuffer(passphrase)) ||
+        (!isInput && cipher != null && !isStringOrBuffer(passphrase))) {
+      throw new ERR_INVALID_OPT_VALUE(option('passphrase', objName),
+                                      passphrase);
+    }
+  }
+
+  return { format, type, cipher, passphrase };
+}
+
+/*
```

Comment:
This doesn't disturb me, but don't we usually use `// ` comment leaders, even for blocks?